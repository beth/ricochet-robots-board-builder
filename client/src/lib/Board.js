import GLYPH_QUADRANTS from './config/glyphQuadrants';
import * as square from './Square';

/*
  BOARD HELPERS
*/
const determineQuadrant = (size, row, col) => {
  if (row <= size - 1 && col <= size - 1) {
    return 'TL';
  } else if (row <= size - 1) {
    return 'TR';
  } else if (col <= size - 1) {
    return 'BL';
  }
  return 'BR';
};

const glyphAllowedInColorQuadrant = (glyph, color) => GLYPH_QUADRANTS[glyph.getName()] === color;

const isConnector = (board, row, col) => Boolean(board.getIn(['squares', row, col, 'connector']));

const remainingQuadrant = board => board.get('quadrants').reduce((memo, color, quadrant) => (color ? memo : quadrant), null);

/*
  BOARD METHODS
*/
const setQuadrant = (board, quadrant, color) => {
  let newBoard = board;
  if (!board.get('remainingColors').has(color)) {
    return board;
  }
  const previousColor = board.getIn(['quadrants', quadrant]);
  if (previousColor) {
    newBoard = newBoard.update('remainingColors', remainingColors => remainingColors.add(previousColor));
  }
  newBoard = newBoard.setIn(['quadrants', quadrant], color);
  newBoard = newBoard.update('remainingColors', remainingColors => remainingColors.delete(color));
  if (newBoard.get('remainingColors').size === 1) {
    newBoard = setQuadrant(newBoard, remainingQuadrant(newBoard), newBoard.get('remainingColors').first());
  }
  return newBoard;
};

const setGlyph = (board, glyph, row, col) => {
  let newBoard = board;
  const quadrant = determineQuadrant(board.get('size'), row, col);
  if (!glyphAllowedInColorQuadrant(glyph, board.getIn(['quadrants', quadrant])) || isConnector(board, row, col)) {
    return newBoard;
  }
  const currentGlyph = board.getIn(['squares', row, col, 'glyph']);
  if (currentGlyph) {
    newBoard = newBoard.update('remainingGlyphs', remainingGlyphs => remainingGlyphs.add(currentGlyph));
  }
  newBoard = newBoard.updateIn(['squares', row, col], squareToUpdate => square.setGlyph(squareToUpdate, glyph));
  newBoard = newBoard.update('remainingGlyphs', remainingGlyphs => remainingGlyphs.delete(glyph));
  return newBoard;
};

export {
  determineQuadrant,
  glyphAllowedInColorQuadrant,
  isConnector,
  remainingQuadrant,
  setQuadrant,
  setGlyph,
};
