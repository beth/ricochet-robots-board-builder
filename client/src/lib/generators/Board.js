import { fromJS, Set } from 'immutable';
import generateGlyphs from './Glyph';
import generateSquares from './Square';
import COLORS from '../config/colors';

const Board = size => ({
  size,
  squares: generateSquares(size),
  quadrants: {
    TL: null,
    TR: null,
    BL: null,
    BR: null,
  },
});

const generateBoard = (size) => {
  let board = fromJS(Board(size));
  board = board.set('remainingGlyphs', Set(generateGlyphs()));
  board = board.set('remainingColors', Set(COLORS));
  return board;
};

export default generateBoard;
