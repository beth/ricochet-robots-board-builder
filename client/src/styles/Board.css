:local .board {
  display: grid;
  grid-template-rows: repeat(16, 50px);
}
