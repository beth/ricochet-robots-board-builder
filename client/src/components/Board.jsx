import React from 'react';
import PropTypes from 'prop-types';
import Row from './Row';
import styles from '../styles/Board.css';

const Board = props => (
  <div className={styles.board}>
    {props.squares.map(row => <Row squares={row} key={row[0].row} />)}
  </div>
);

Board.propTypes = {
  squares: PropTypes.arrayOf(PropTypes.array).isRequired,
};

export default Board;
