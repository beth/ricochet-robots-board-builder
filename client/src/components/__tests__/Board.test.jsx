import React from 'react';
import renderer from 'react-test-renderer';
import Board from '../Board';
import BoardConstructor from '../../lib/Board';

describe('Board Component', () => {
  it('renders correctly', () => {
    const board = new BoardConstructor(16);
    const domTree = renderer.create(<Board squares={board.squares} />).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});
