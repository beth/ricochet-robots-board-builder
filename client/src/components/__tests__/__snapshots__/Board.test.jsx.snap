// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Component renders correctly 1`] = `
<div
  className="board"
>
  <div
    className="row"
  >
    <div
      className="top left square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top square"
    />
    <div
      className="top right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="top left square connector"
    />
    <div
      className="top right square connector"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="bottom left square connector"
    />
    <div
      className="bottom right square connector"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="left square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="square"
    />
    <div
      className="right square"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="bottom left square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom square"
    />
    <div
      className="bottom right square"
    />
  </div>
</div>
`;
